### DISCUSSION TITLE
sRGB to CMYK color conversion (Not Image) with icc profile using `xc:` and `-format`

### DISCUSSION THREAD

**Discussion by necessarylion** [dlemstra/magick-wasm#191] [Q&A]
Hi, I have a magick command that can convert sRGB color to CMYK color using icc profile.
I am looking for alternative solution using this `magick-wasm` library.

```ts
magick xc:"srgb(255,255,255,255)" -profile sRGB.icc -profile CoatedFOGRA39.icc -colorspace CMYK  -format '%[pixel:u]' info:

OUTPUT - cmyk(0,0.622568,0.0933852,0)
```

References
- https://stackoverflow.com/a/47023522/9773835
- https://stackoverflow.com/a/48191994/9773835

**Comment by dlemstra** [comment]
What have you tried so far and where are you stuck?

**Reply by necessarylion** [reply]
I haven't tried any solutions yet, as the package might not be designed for this purpose.

I'm looking for something like:

```ts
image.xc("srgb(255,255,255,255)")
image.format("%[pixel:u]", ":info") 
```

or perhaps a function to execute custom commands.
or perhaps color converter function in `MagickColor` class.

I try to use [jsColorEngine](https://github.com/glennwilton/jsColorEngine), but the output doesn't quite match what I get with ImageMagick.

Currently, to convert colors, I rely on a backend service where ImageMagick is installed. I'm searching for a solution to perform this conversion directly in the browser using JavaScript.

**Reply by glennwilton** [reply]
I'm the author of jsColorEngine, recently just released some big updates https://github.com/glennwilton/jsColorEngine  if you still have a need for javascript icc colour conversions.

**Comment by dlemstra** [comment]
`xc:"srgb(255,255,255,255)"` should be the name of the image e.g. : `new MagickImage('xc:srgb(255,255,255,255)')` but the `format` argument is not yet available in the api. I can add this for you.

**Reply by necessarylion** [reply]
Didn't know I can just do that.😃 
Yeah It work! and it would be great if you can add the feature for the format (`%[pixel:u]`). 

BTW, thank you for this amazing library.

**Reply by dlemstra** [reply]
This shouldn't be that difficult to add. I now also wonder if the `transformColorSpace` method should change the `colorSpace` of the image to the `colorSpace` of the target profile. I was creating a unit test and after setting the profile I also had to change the `colorSpace` before the correct value would be returned in the new `formatExpression` method. Your thoughts?

**Reply by necessarylion** [reply]
I'm not entirely sure how `transformColorSpace` works, but I believe it would be better to follow the standard approach of the parent ImageMagick command:
```js
magick -profile input.icc -profile output.icc
```

For example, when transforming a color space, we typically use profiles twice and setting color space separately. Therefore, it should look like this:

```ts
image.profile(input);
image.profile(output);
image.colorSpace = ColorSpace.CMYK;
```

I use the `gm` npm package (https://www.npmjs.com/package/gm) on the server side for my app. And Its design also requires handling profile twice and setting the colorspace have to done separately.

```ts
const imgMagick = gm.subClass({ imageMagick: '7+' })
imgMagick(bytes)
  .profile(inputProfile)
  .profile(outputProfile)
  .colorspace('CMYK')
  .background('#FFFFFF')
  .flatten()
  .quality(100)
  .setFormat('jpg')
  .trim()
  .toBuffer((err, buffer) => {
    if (err) rejects(err)
    if (buffer) resolve(Uint8Array.from(buffer))
  })
```

**Reply by dlemstra** [reply]
I checked this with my Magick.NET library and I want it to behave in the same way.

I just pushed a patch to remove the `name` argument of `ColorSpace` constructor. Only `icc` will be used in the future. The `transformColorSpace` should change the `colorSpace` of an image using the specified color profile. In the next release you no longer need to do `image.colorSpace = ColorSpace.CMYK;`. You can still do this but you don't need to do it anymore.

**Reply by dlemstra** [reply]
I also pushed a patch to add `formatExpression` to `MagickImage`, this will become available in the next release.

**Reply by necessarylion** [reply]
Yeah. It Make scene. Thanks for the adding this feature.

**Comment by necessarylion** [comment]
Hi, @dlemstra 
While I was testing newly added feature `formatExpression`. I found out that the result did not match exactly (but correct).

I convert this `srgb(255,255,255,255)`
and in CLI command got this `cmyk(0,0.622568,0.0933852,0)`.
But in magick-wasm I got this `cmyk(0,1,0,0)`.

So basically, in wasm version, values are parsed into integer.

So not sure if we can use `-precision` for it. I try to check if there is precision function, unfortunately there is not.

**Reply by dlemstra** [reply]
Are you using the Q8 version on the command line? Because this library uses the Q8 version of ImageMagick. I have not yet made getting or setting the precision available in the api because nobody requested this. Also not sure if we should add this.

**Reply by necessarylion** [reply]
Seem like Q16.

```
Version: ImageMagick 7.1.1-38 Q16-HDRI aarch64 22398 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP
Delegates (built-in): bzlib fontconfig freetype gslib heic jng jp2 jpeg jxl lcms lqr ltdl lzma openexr png ps raw tiff webp xml zlib zstd
Compiler: gcc (4.2)
```

Yeah. no need to add `-precision` for now.

**Reply by dlemstra** [reply]
This is indeed happening because of the Q8 depth. And I also though I did not need to add the `ColorTransformMode` but it turns out I need to do that to keep the results the same as the Magick.NET library. This will also become available in the next release.

**Reply by necessarylion** [reply]
Really appreciate your support. Looking forward for the release.

**Reply by dlemstra** [reply]
Just to clarify the next release will still show it as integers instead of floating point numbers because this wasm library has a quantum depth of 8 bit and not 16 bit.

**Reply by necessarylion** [reply]
It should be fine. Thanks.
